### ISSUE TITLE
using a post callback route as shown in the readme doesn't work

### ISSUE THREAD

**Issue by tylercal** [omniauth/omniauth#1028]
The Readme gives the example of how to configure your rails app by adding
```ruby
post '/auth/:provider/callback', to: 'sessions#create'
```
But this is the route that is called back to _after_ you've been authenticated with the provider. This will be a `GET` method from Google (and I assume most others).

### Configuration

- Provider Gem: `omniauth-2.0.1`
- Ruby Version: `2.6.6`
- Framework: `rails`
- Platform: `macOS`

### Expected Behavior

When following the documentation in the README on how to configure your `routes.rb`, omniauth should work as expected.

### Actual Behavior

Auth requests fail with `No route matches [GET] "/auth/google_oauth2/callback"`

### Steps to Reproduce

Configure a new rails app with `omniauth`, `omniauth-rails_csrf_protection`, and `omniauth-google-oauth2`
add the following route: `post '/auth/:provider/callback', to: 'sessions#create`
attempt to auth via Google

**Comment by VedaRamaiah** [comment]
Have you registered the callback on the provider?

**Comment by tylercal** [comment]
Yes, this was a previously working provider. Google doesn't offer any option (that I can see) to set the method that should be used during callback in the redirect_uri. There is also nothing specified in [their documentation](https://developers.google.com/identity/protocols/oauth2/openid-connect).

**Comment by BobbyMcWho** [comment]
If your provider issues a GET callback, then use the GET callback route.
get '/auth/:provider/callback', to: 'sessions#create

The POST changes in OmniAuth 2.0 are for the request phase, and should not affect the callback phase.

**Comment by tylercal** [comment]
Are there providers that have a POST callback? Google certainly doesn’t and unless Facebook has recently changed, that is a GET as well. My suggestion is that the readme should be updated to reflect the most common use cases.

**Comment by rosewcs345** [comment]
Either that, or try this:
 match '/auth/:provider/callback', :to => 'sessions#create', :via => [:get, :post]

**Comment by joshIsCoding** [comment]
Worth noting as well that mock callback requests, as generated in integration tests when running in test mode, also use GET with no obvious way of changing to POST, so such tests fail when specifying a POST-only callback route. Is there a way to get POST callbacks in test mode?

**Comment by deniciocode** [comment]
I have a kinda stupid question on this topic. Where do I see which route is available?

**Comment by tylercal** [comment]
There are [several options](https://guides.rubyonrails.org/routing.html#inspecting-and-testing-routes) to see what routes are available for a rails application. But in summary
* visit http://localhost:3000/rails/info/routes
* execute `bin/rails routes` on the terminal in your project's root directory

**Comment by deniciocode** [comment]
Maybe I should have asked in a more descriptive way. `rails routes` shows me routes, but does not show me the routes for `omniauth`. When I plugin in `omniauth-steam` I am able to `POST` to `/auth/steam` but this route is not show when I am using `rails route`. 

And the question is, how do I get the list of available routes?

**Comment by tylercal** [comment]
Got it, hopefully this explains it a bit better

See the [Integrating omniauth into your application](https://github.com/omniauth/omniauth#integrating-omniauth-into-your-application). The route you're adding to your routes.rb is 
```ruby
post '/auth/:provider/callback', to: 'sessions#create'
```

In this case the `:provider` is a [dynamic segment](https://guides.rubyonrails.org/routing.html#dynamic-segments) that makes whatever appears after the `/auth/` and before the `/callback` available as a parameter to omniauth. Omniauth uses this parameter to lookup the provider you're trying to auth with.

This typically appears in an initializer with a line like
```ruby
provider :steam, "app-id", "secret" 
``` 
but you can customize it by adding an additional entry to the hash like 
```ruby
provider :steam, "app-id", "secret" , name: gamer
```
Which would then create an auth route for `auth/gamer/callback`. This comes in handy when you want to use the same provider in multiple ways.

**Comment by tylercal** [comment]
This issue was addressed in commit 898be28

**Comment by jon-sully** [comment]
That commit didn't actually change the problem — the README instructs us to put a `post` route into our routes file but the actual callback workflow yields a GET 🤔
